Part 4: Refactor implementation of TemporalPlainDate/prototype/add

ptomato opened the pull request:
#### 707926f7f3389c6ad904c8c2f9d4b13169c63838
<pre>
Part 4: Refactor implementation of TemporalPlainDate/prototype/add
<a href="https://bugs.webkit.org/show_bug.cgi?id=223166">https://bugs.webkit.org/show_bug.cgi?id=223166</a>

Reviewed by Sosuke Suzuki.

Add a new operation from the spec, TemporalCalendar::addDurationToDate(),
and use it in the implementation of `add`.

This is a refactoring and shouldn&apos;t change anything.

(For reference, modifications by Philip Chimento to address review
comments:)

* Remove ISO8601::isValidISODate() and ISO8601::createISODateRecord()
  since they are not used yet in this PR.
* Expose JSC::checkedCastDoubleToInt128 in ISO8601.h header.
* Replace C-style double-to-Int128 casts with checked casts.
* Assert that TemporalDuration::timeDurationFromComponents and
  add24HourDaysToTimeDuration do not overflow. (The limits on individual
  Duration components should ensure that.)

Canonical link: <a href="https://commits.webkit.org/299436@main">https://commits.webkit.org/299436@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0bd083b51e95e4a1fb3d37b55556fc2b8a9422e5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119020 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38701 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29352 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125225 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71083 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39397 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47283 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/90374 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59824 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121973 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31412 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106704 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70832 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/30475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24820 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68878 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/111138 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100859 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25006 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128261 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/117535 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45927 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34703 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98990 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46294 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102924 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98770 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25102 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44228 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22227 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/42506 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45797 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51475 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/146233 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45263 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37618 "Found 11 new JSC stress test failures: wasm.yaml/wasm/lowExecutableMemory/executable-memory-oom.js.default-wasm, wasm.yaml/wasm/stress/cc-int-to-int-cross-module-with-exception.js.default-wasm, wasm.yaml/wasm/stress/cc-int-to-int-jit-to-llint.js.default-wasm, wasm.yaml/wasm/stress/wasm-js-string-builtins.js.default-wasm, wasm.yaml/wasm/stress/wasm-js-string-builtins.js.wasm-bbq, wasm.yaml/wasm/stress/wasm-js-string-builtins.js.wasm-bbq-no-consts, wasm.yaml/wasm/stress/wasm-js-string-builtins.js.wasm-collect-continuously, wasm.yaml/wasm/stress/wasm-js-string-builtins.js.wasm-eager, wasm.yaml/wasm/stress/wasm-js-string-builtins.js.wasm-eager-jettison, wasm.yaml/wasm/stress/wasm-js-string-builtins.js.wasm-no-cjit ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48609 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46949 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on Source/JavaScriptCore/runtime/TemporalDuration.cpp:
```diff
@@ -212,6 +212,21 @@ static double totalNanoseconds(ISO8601::Duration& duration)
+
+    Int128 daysInNanoseconds = static_cast<Int128>(days) * ISO8601::ExactTime::nsPerDay;
+    Int128 result = d + daysInNanoseconds;
+    if (absInt128(result) > ISO8601::InternalDuration::maxTimeDuration) {
```
Can you add `[[unlikely]]` ?

sosukesuzuki commented on Source/JavaScriptCore/runtime/TemporalDuration.cpp:
```diff
@@ -304,6 +319,43 @@ static TemporalUnit largestSubduration(const ISO8601::Duration& duration)
+    Int128 sec = static_cast<Int128>(seconds) + min * 60;
+    Int128 millis = static_cast<Int128>(milliseconds) + sec * 1000;
+    Int128 micros = static_cast<Int128>(microseconds) + millis * 1000;
+    Int128 nanos = static_cast<Int128>(nanoseconds) + micros * 1000;
```
I think we should use `checkedCastDoubleToInt128` to safely add, similar to `Duration::totalNanoseconds`

ptomato replied:
Hmm. Looks like checked multiply with int128 doesn't work on armv7. I'll make some inquiries and see if I can fix this on Wednesday, but your advice would be appreciated as well.

sosukesuzuki replied:
Hmm, gcc building an ARMv7 port might not be able to perform multiplication between Int128 and int32. As totalNanoseconds is doing, changing it to multiplication between Int128 and Int128 might make it buildable?

ptomato replied:
Thanks for the suggestion. I'll check if this current revision builds.

sosukesuzuki commented on Source/JavaScriptCore/runtime/ISO8601.h:
```diff
@@ -321,6 +321,8 @@ String monthCode(uint32_t);
 
 bool isValidDuration(const Duration&);
+bool isValidISODate(double, double, double);
+PlainDate createISODateRecord(double, double, double);
```
This function is not used. Is it correct to include it in this PR?

ptomato replied:
I think Tim intentionally included it (see commit message) because it was part of a larger commit that he split up, but I can remove it if that's better.

sosukesuzuki commented on Source/JavaScriptCore/runtime/TemporalDuration.cpp:
```diff
@@ -304,6 +320,44 @@ static TemporalUnit largestSubduration(const ISO8601::Duration& duration)
+    CheckedInt128 millis = checkedCastDoubleToInt128(milliseconds) + sec * Int128 { 1000 };
+    CheckedInt128 micros = checkedCastDoubleToInt128(microseconds) + millis * Int128 { 1000 };
+    CheckedInt128 nanos = checkedCastDoubleToInt128(nanoseconds) + micros * Int128 { 1000 };
+    ASSERT(!nanos.hasOverflowed());
+    ASSERT(absInt128(nanos) <= ISO8601::InternalDuration::maxTimeDuration);
```
Are these two assertions guaranteed somewhere? If they're not guaranteed, I think we need to throw a `RangeError` or something here instead of using `ASSERT`

ptomato replied:
I believe they are guaranteed, so I'll add some clarifying comments.

sosukesuzuki reviewed (approved):
r=me

catamorphism committed:
Part 4: Refactor implementation of TemporalPlainDate/prototype/add
https://bugs.webkit.org/show_bug.cgi?id=223166

Reviewed by Sosuke Suzuki.

Add a new operation from the spec, TemporalCalendar::addDurationToDate(),
and use it in the implementation of `add`.

This is a refactoring and shouldn't change anything.

(For reference, modifications by Philip Chimento to address review
comments:)

* Remove ISO8601::isValidISODate() and ISO8601::createISODateRecord()
  since they are not used yet in this PR.
* Expose JSC::checkedCastDoubleToInt128 in ISO8601.h header.
* Replace C-style double-to-Int128 casts with checked casts.
* Assert that TemporalDuration::timeDurationFromComponents and
  add24HourDaysToTimeDuration do not overflow. (The limits on individual
  Duration components should ensure that.)

Canonical link: https://commits.webkit.org/299436@main
